src/main/java/com/mongodb/kafka/connect/source/schema/BsonDocumentToSchema.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.kafka.connect.source.schema;
+
+import static java.lang.String.format;
+
+import java.util.List;
+import java.util.Objects;
+
+import org.apache.kafka.connect.data.Decimal;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.SchemaBuilder;
+import org.apache.kafka.connect.data.Timestamp;
+
+import org.bson.BsonValue;
+
+public final class BsonDocumentToSchema {
+
+  private static final Schema DEFAULT_INFER_SCHEMA_TYPE = Schema.STRING_SCHEMA;
+  public static final String SCHEMA_NAME_TEMPLATE = "inferred_name_%s";
+
+  public static Schema inferSchema(final BsonValue bsonValue) {
+    switch (bsonValue.getBsonType()) {
+      case BOOLEAN:
+        return Schema.BOOLEAN_SCHEMA;
+      case INT32:
+        return Schema.INT32_SCHEMA;
+      case INT64:
+        return Schema.INT64_SCHEMA;
+      case DOUBLE:
+        return Schema.FLOAT64_SCHEMA;
+      case DECIMAL128:
+        return Decimal.schema(bsonValue.asDecimal128().getValue().bigDecimalValue().scale());
+      case DATE_TIME:
+      case TIMESTAMP:
+        return Timestamp.SCHEMA;
+      case DOCUMENT:
+        SchemaBuilder builder = SchemaBuilder.struct();
+        bsonValue.asDocument().forEach((k, v) -> builder.field(k, inferSchema(v)));
+        builder.name(generateName(builder));
+        return builder.build();
+      case ARRAY:
+        List<BsonValue> values = bsonValue.asArray().getValues();
+        Schema firstItemSchema =
+            values.isEmpty() ? DEFAULT_INFER_SCHEMA_TYPE : inferSchema(values.get(0));
+        if (values.isEmpty()
+            || values.stream().anyMatch(bv -> !Objects.equals(inferSchema(bv), firstItemSchema))) {
+          return SchemaBuilder.array(DEFAULT_INFER_SCHEMA_TYPE).build();
+        }
+        return SchemaBuilder.array(inferSchema(bsonValue.asArray().getValues().get(0))).build();
+      case BINARY:
+        return Schema.BYTES_SCHEMA;
+      case SYMBOL:
+      case STRING:
+        return Schema.STRING_SCHEMA;
+      case OBJECT_ID:
+      case NULL:
```

Comment:
Why does null map to string?

Reply:
Because there is no null type in the Kafka Schema

Reply:
Does this mean the connector will convert `null` to `"null"`?  Or to `""`?

Reply:
It will be whatever the string / json formatter version of null outputs. By default "null"

Reply:
I'm wondering whether it would be better to leave the field out of the schema entirely if the value is null.  What do you think?

Reply:
I think its an extremely rare case, but it is potentially a value so should be included to at least indicate the field exists.

Reply:
Huh I thought this applied to any case like

   {
     "foo": null,
     ...
   }

which isn't rare.  Is that incorrect?

Reply:
I think having null values in documents is a less common pattern, probably not extremely rare though!

I've updated the code, so when inferring the schema I've made BsonNull to have an optional string schema and now set the value null.

Reply:
Agreed that `null` is less common, but not rare either.

It still seems to me like leaving it out entirely is more accurate, since in most query cases in MongoDB (`$exist` being the most prominent exception to the rule) `foo: null` is equivalent to `foo` being absent from the document.  And we have no idea whether non-null values for `foo` are likely to be string rather than any other type.

Reply:
The purpose of the schema is to describe the values in the document. For that a `null` value is different to non-existing value.

Reply:
I see.  So pulling it out of the schema would require pulling it out of the source document as well, which we should not do.  Given than, I'm not sure what else we can do about the impedance mismatch between BSON and schema, beyond documenting the behavior.